src/main/kotlin/it/unibo/protelis/web/execution/simulated/SimulatedProtelisEngine.kt

```diff
@@ -9,71 +9,74 @@ import it.unibo.alchemist.model.implementations.positions.Euclidean2DPosition
-
-  init {
-    logger.debug("Simulation container created")
-  }
-
-  constructor(sourceCode: String, monitor: ProtelisObserver) : this() {
-    GlobalScope.launch { setupAwait(sourceCode, monitor) }.asCompletableFuture().join()
-  }
+  private var alchemistEngine: Simulation<Any, Euclidean2DPosition>? = null
 
   override suspend fun setupAwait(sourceCode: String, monitor: ProtelisObserver) {
-    if (alchemistEngine.get() != null) stopAwait()
+    if (alchemistEngine != null) stopAwait()
     val sim = setupSimulationAsync().await()
-    alchemistEngine.set(sim)
+    alchemistEngine = sim
     setCode(sourceCode)
-    alchemistEngine.get()
+    alchemistEngine
       ?.addOutputMonitor(ProtelisUpdateAdapter(monitor))
       ?: throw IllegalStateException("Simulation was not set up correctly")
-    alchemistEngine.get()
+    alchemistEngine
       ?.awaitFor(Status.INIT)
       ?: throw IllegalStateException("Simulation was not set up correctly")
-    GlobalScope.launch { sim.run() }
+    GlobalScope.launch { alchemistEngine?.run() ?: throw IllegalStateException() }
     logger.debug("Simulation Engine set up correctly")
   }
 
   override suspend fun startAwait() {
-    checkNotNull(alchemistEngine.get())
-    alchemistEngine.get()?.play()
-    alchemistEngine.get()?.awaitFor(Status.RUNNING) // TODO: should I wait RUNNING instead?
+    checkNotNull(alchemistEngine)
+    alchemistEngine
+      ?.play()
+      ?: throw IllegalStateException()
+    alchemistEngine
+      ?.awaitFor(Status.RUNNING)
+      ?: throw IllegalStateException()
     logger.debug("Simulation started")
   }
 
   override suspend fun stopAwait() {
-    checkNotNull(alchemistEngine.get())
-    alchemistEngine.get()?.terminate()
-    alchemistEngine.get()?.awaitFor(Status.TERMINATED)
-    alchemistEngine.set(null)
+    checkNotNull(alchemistEngine)
+    alchemistEngine
+      ?.terminate()
+      ?: throw IllegalStateException()
+    alchemistEngine
+      ?.awaitFor(Status.TERMINATED)
+      ?: throw IllegalStateException()
+    alchemistEngine = null
     logger.debug("Simulation stopped and deleted")
   }
 
   /** Asynchronously build a new simulation from a loader of YAML files. */
-  private fun setupSimulationAsync(): Deferred<Engine<Any /* TODO */, Euclidean2DPosition>> = GlobalScope.async {
-    val loader = YamlLoader(this.javaClass.classLoader.getResourceAsStream("simulation.yml"))
-    loader.getDefault<Any /* TODO */, Euclidean2DPosition>()
-    Engine<Any /* TODO */, Euclidean2DPosition>(
-      loader.getDefault(),
+  private fun setupSimulationAsync(): Deferred<Engine<Any, Euclidean2DPosition>> = GlobalScope.async {
+    Engine<Any, Euclidean2DPosition>(
+      YamlLoader(this.javaClass.classLoader.getResourceAsStream("simulation.yml")).getDefault(),
       Long.MAX_VALUE,
       DoubleTime(Double.POSITIVE_INFINITY)
     )
   }
 
+  /** Build a new simulation from a loader of YAML files. */
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [CommentOverPrivateFunction - [setupSimulation] at /src/src/main/kotlin/it/unibo/protelis/web/execution/simulated/SimulatedProtelisEngine.kt:70:3](https://app.codacy.com/manual/NiccoMlt/Protelis-Web/pullRequest?prid=5040950)